### PR TITLE
Add a phase to perform functional to inplace operator conversion

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57477 [JIT] Add a phase to perform functional to inplace operator conversion**

Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

### CODE REVIEW

File: test/jit/test_convert_activation.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        self.assertEqual(model(inp), frozen_model(inp))
+
+class TestInplaceToFunctionalActivation(JitTestCase):
+    def test_inplace_to_functional_activation(self):
+        for activation in [
+            torch.nn.functional.hardsigmoid,
+            torch.nn.functional.hardtanh,
+            torch.nn.functional.hardswish,
+            torch.nn.functional.relu,
+            torch.nn.functional.relu6,
```
Comment (penguinwu) [inline test/jit/test_convert_activation.py:59]: why is `sigmoid` and `tanh` included in the `TestFunctionalToInplaceActivation` but not `TestInplaceToFunctionalActivation`?

If we are supposed to test exactly the same set of activation functions for both tests, then perhaps to define a global `ActivationList` that is used by both tests would avoid such asymmtry.

File: torch/csrc/jit/passes/convert_activation.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    // Use y = relu(x) as an example:
+    // If y and x have different type, skip the conversion
+    if (input->type() != output->type()) {
+      continue;
+    }
+
+    // If x is an input parameter or alias with an input parameter, skip the
+    // conversion
+    if (input->node()->kind() == prim::Param ||
```
Comment (penguinwu) [inline torch/csrc/jit/passes/convert_activation.cpp:54]: what if the input is a weight tensor? Basically, in model execution, are there other types of tensors (other than param) whose live-range exceeds the current inference execution?

I think the criteria here is to make sure that the input is a last-use. There are tensors who are alive throughout the inference (e.g., graph inputs or weights).

File: torch/csrc/jit/passes/convert_activation.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    if (functional_activation_ops_.count(node->kind()) == 0) {
+      continue;
+    }
+
+    Value* input = node->inputs().at(0);
+    Value* output = node->output();
+
+    // Use y = relu(x) as an example:
+    // If y and x have different type, skip the conversion
+    if (input->type() != output->type()) {
```
Comment (penguinwu) [inline torch/csrc/jit/passes/convert_activation.cpp:48]: What if the input and output sizes differ?

File: torch/csrc/jit/passes/convert_activation.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    if (functional_activation_ops_.count(node->kind()) == 0) {
+      continue;
+    }
+
+    Value* input = node->inputs().at(0);
+    Value* output = node->output();
+
+    // Use y = relu(x) as an example:
```
Comment (penguinwu) [inline torch/csrc/jit/passes/convert_activation.cpp:46]: The key part of the algorithm is to decide which activation functions can be rewritten to the in-place form. Could you carve out that part of the logic into a separate helper function and describe the algorithm in the comments? I feel that the current criteria is safe but not optimal and may be improved later and labeling this part of the algorithm separately may make it easier for future update

File: torch/csrc/jit/passes/convert_activation.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/utils/memory.h>
+
+namespace torch {
+namespace jit {
+
+class FunctionalActivationRewriter {
+ public:
```
Comment (penguinwu) [inline torch/csrc/jit/passes/convert_activation.h:13]: Please add comments to describe what the passes do

**Review by penguinwu — CHANGES_REQUESTED** [review]
I have a couple of questions regarding the algorithm that checks whether a functional op can be replaced by functional op. Please check those.

File: torch/csrc/jit/passes/convert_activation.cpp (hunk 3 above)
Comment (eellison) [inline torch/csrc/jit/passes/convert_activation.cpp:48]: So, this check is getting at the fact that we can't inplace a node if the output dtype is separate from the input dtype. Could you comment that ? Shape isn't necessary for activation since the size will always remain unchanged. 

If we're going to use `input->type() != output->type()` to verify that the output dtype equals the input dtype, we also need to check that the dtype is actually specified on the types. It is not guaranteed to be. 

I would maintain a list of activations you know do not type promote (relu for instance), and then for the ones not in that list that do type promote (sigmoid), check that the dtype is specified on both input and output and they are equal.

File: torch/csrc/jit/passes/convert_activation.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    // If x has more than one use, skip the converson.
+    // TODO: Use more accurate analysis to relax this restriction.
+    if (input->uses().size() != 1) {
+      continue;
+    }
+
+    changed = true;
+    Node* inplace_node = node->replaceWithNewSymbol(
+        Symbol::fromQualString(node->schema().name() + "_"));
+    inplace_node->output()->replaceAllUsesWith(input);
```
Comment (eellison) [inline torch/csrc/jit/passes/convert_activation.cpp:68]: Even through using the input or output are functionally equivalent, it is more idiomatic to use the output.

File: torch/csrc/jit/passes/convert_activation.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    getOrCreateAliasDb()->replaceWithNewValue(output, input);
```
Comment (eellison) [inline torch/csrc/jit/passes/convert_activation.cpp:69]: Can we do `getOrCreateAliasDb()->replaceWithNewValue(node->output, inplace_node->output());` ?

**Review by eellison — COMMENTED** [review]
Looks good, I think we need to update the logic around when it is safe to reinplace nodes around type promotion.

File: test/jit/test_convert_activation.py (hunk 1 above)
Comment (desertfire) [reply test/jit/test_convert_activation.py:59]: torch.sigmoid or torch.tanh use a different syntax for its inplace version. I will rewrite this test.

File: torch/csrc/jit/passes/convert_activation.cpp (hunk 6 above)
Comment (desertfire) [reply torch/csrc/jit/passes/convert_activation.cpp:68]: > Even through using the input or output are functionally equivalent, it is more idiomatic to use the output.

Are you suggesting `inplace_node->input()->replaceAllUsesWith(output)`? I personally found that a little counter-intuitive.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

File: torch/csrc/jit/passes/restore_mutation.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    return false;
+  }
+  auto inputDtype = input->type()->cast<TensorType>()->scalarType();
+  auto outputDtype = output->type()->cast<TensorType>()->scalarType();
+
+  // if (check_shape) {
+  //  return false;
+  //}
+
+  // In general, we don't need to check shape as activation ops are
```
Comment (eellison) [inline torch/csrc/jit/passes/restore_mutation.cpp:68]: We're not checking that there's an equivalent inplace schema for the node when making this transformation, which we know for activations but we don't know if we're making this a public exposed api

**Review by eellison — COMMENTED** [review]
Looking good! One comment is that i'm not really sure if opening this up as a public API makes sense... I don't know what the intended use case is and the extra abstractions are making the pass more complicated. If we are opening this up as a public API we need to make sure that the conditions are correct. 

The other comment is that the operators we can reinplace are the operators that dont type promote or dont accept integer inputs. if you look at [this script](https://gist.github.com/eellison/aa67f0f13d3205538e77acfb22dd397d) I think it's a greater set than you have. We might consider checking in some version of this script as a test so that if integer inputs are added or type promotion is added to these operators the test will fail. 

`selu
hardsigmoid
hardtanh
hardswish
rrelu_
elu
relu
leaky_relu
celu
relu6
silu`

File: torch/csrc/jit/passes/restore_mutation.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+// A common application scenario is to apply InplaceToFunctionalActivation
+// before some JIT optimization passes, so that those passes are less
+// constrained by in-place ops. After those passes are done, we can call
+// FunctionalToInplaceActivation to recover in-place activation ops,
+// so that we won't lose the performance benefit coming from memory reduction.
+
+// Replaces in-place aten activation ops with their functional equivalents
+TORCH_API bool InplaceToFunctionalActivation(
```
[the remark this answers is not in the document]
Comment (desertfire) [reply torch/csrc/jit/passes/restore_mutation.h:50]: Initially I was thinking to put it in remove_mutation.cpp, but that requires we expose the activation op set as global. I don't see many other passes are exposing this kind of global data, but if that is acceptable, I can change the code.

File: test/jit/test_convert_activation.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        frozen_model = torch.jit.freeze(torch.jit.script(model.eval()))
+        N, C, H, W, = 10, 3, 224, 224
+        inp = torch.randn(N, C, H, W)
+        self.run_pass('functional_to_inplace_activation', frozen_model.graph)
+        self.assertEqual(model(inp), frozen_model(inp))
+
+
+class TestInplaceToFunctionalActivation(JitTestCase):
+    def test_inplace_to_functional_activation(self):
+        for activation in [
```
[the remark this answers is not in the document]
Comment (desertfire) [reply test/jit/test_convert_activation.py:90]: Because the activation syntax are different for the two groups, `_` vs `inplace=True`

File: torch/csrc/jit/passes/restore_mutation.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (desertfire) [reply torch/csrc/jit/passes/restore_mutation.h:50]: We probably need a centralized location to store the activation op sets, or is it possible to add attributes (e.g. is_elementwise, is_type_promotable, ..) for our builtin ops?

File: torch/csrc/jit/passes/restore_mutation.cpp (hunk 8 above)
Comment (desertfire) [reply torch/csrc/jit/passes/restore_mutation.cpp:68]: I will add a schema checking.

File: torch/csrc/jit/passes/restore_mutation.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+#include <c10/util/Exception.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/utils/memory.h>
+
+namespace torch {
+namespace jit {
+
+struct FunctionalToInplaceRewriterFlags {
```
[the remark this answers is not in the document]
Comment (desertfire) [reply torch/csrc/jit/passes/restore_mutation.h:12]: I was thinking to use this to later support more builtin ops beyond activation ops. For example, if later on we want to support add ->add_ , we want to check the shape after the shape analysis is more mature, but we still won't need to check shape for element-wise ops. I guess a cleaner way to implement this is to associate this kind of information to each group of ops.

[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/restore_mutation.h:12]: Implementing it for add would add a dependency on dtype/shape analysis, I don't think this API would be part of it.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

File: test/jit/test_convert_activation.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        # restore_mutation.h contains a mapping from activation operators
+        # to whether they allow type conversion. Use this checking to
+        # guard the mapping, and if any later change breaks the assumption
+        # we need to update the mapping correspondingly.
+        for activation, dtype in product(activations, dtypes):
+            inp = torch.normal(0, 5, size=(4, 4)).to(dtype)
+            try:
+                out = activation(inp)
+                if dtype != out.dtype:
+                    exit()
```
Comment (eellison) [inline test/jit/test_convert_activation.py:62]: nit: i'm unfamiliar with exit, never seen it before...

File: test/jit/test_convert_activation.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        dtypes = [
+            torch.bool,
+            torch.int8,
+            torch.int16,
+            torch.int32,
+            torch.int64,
+            torch.float32,
+            torch.float64,
+        ]
+        # restore_mutation.h contains a mapping from activation operators
```
Comment (eellison) [inline test/jit/test_convert_activation.py:53]: nice !

**Review by eellison — APPROVED** [review]
looks great 😍 😍 😍 great job!!

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).
